pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,367 @@
+			}
+
+			blockOidsData, err := bfs.get(blockOidsObject[0].String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to fetch 'block_oids' object %d", startIndex), zap.Error(err))
+				return err
+			}
+			blockOIDs, err := parseBlockOIDs(blockOidsData)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to parse 'block_oids' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			if len(blockOIDs) == 0 {
+				bfs.log.Info(fmt.Sprintf("No block found with index %d, stopping.", startIndex))
+				return errors.New("no block found")
+			}
+
+			err = bfs.fetchAndProcessBlocks(blockOIDs[skip:])
+			if err != nil {
+				return err
+			}
+			startIndex++
+			skip = 0
+		}
+	}
+}
+
+func (bfs *Service) processBlock(rc io.ReadCloser) error {
+	b := block.New(bfs.stateRootInHeader)
+	r := gio.NewBinReaderFromIO(rc)
+	b.DecodeBinary(r)
+	err := bfs.queue.PutBlock(b)
+	if err != nil {
+		bfs.log.Error("failed to queue block", zap.Error(err))
+		return err
+	}
+	return nil
+}
+
+func (bfs *Service) GetHeaders() error {
+	startIndex := bfs.chain.HeaderHeight() / 2000
+
+	for {
+		select {
+		case <-bfs.quit:
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(bfs.cfg.HeaderAttribute, fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter(bfs.cfg.HeaderAttribute, fmt.Sprintf("%d", startIndex), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			headerOids, err := bfs.search(prm)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to find 'index_header' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			if len(headerOids) == 0 {
+				bfs.log.Info(fmt.Sprintf("No 'index_header' object found with index %d, stopping.", startIndex))
+				return errors.New("no 'index_header' object found")
+			}
+
+			rc, err := bfs.get(headerOids[0].String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to fetch 'index_header' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			err = bfs.processHeaders(rc)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to process headers for index %d", startIndex), zap.Error(err))
+				return err
+			}
+			startIndex++
+		}
+	}
+}
+
+func (bfs *Service) processHeaders(rc io.ReadCloser) error {
+	var resHeader payload.Headers
+	br := gio.NewBinReaderFromIO(rc)
+	resHeader.DecodeBinary(br)
+	err := bfs.chain.AddHeaders(resHeader.Hdrs...)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (bfs *Service) get(oid string) (io.ReadCloser, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+	defer cancel()
+	u, err := url.Parse(fmt.Sprintf("neofs:%s/%s", bfs.cfg.ContainerID, oid))
+	if err != nil {
+		return nil, err
+	}
+	rc, err := neofs.GetWithClient(ctx, bfs.client, bfs.account.PrivateKey(), u, bfs.cfg.Addresses[0])
+	defer rc.Close()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Have you tested the updated version? This code won't work, because you close `rc` in deferred statement, it will be executed right before the method returns. As a result, no one will be able to read something from `rc`. https://go.dev/tour/flowcontrol/12